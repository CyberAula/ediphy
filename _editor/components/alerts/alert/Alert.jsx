import React, { Component } from 'react';
import { Button, Modal } from 'react-bootstrap';

export default class Alert extends Component {
    constructor(props) {
        super(props);
    }
    render() {
        return(
            <Modal className={this.props.className || 'popupAlert pageModal'} backdrop={this.props.backdrop === undefined ? true : this.props.backdrop} show={this.props.show} onHide={e=>{this.props.onClose(e);}}>
                {this.props.hasHeader ?
                    (this.props.closeButton ?
                        (<Modal.Header closeButton>
                            {this.props.title}
                        </Modal.Header>) :
                        (<Modal.Header>
                            {this.props.title}
                        </Modal.Header>)) :
                    (null)}
                <Modal.Body>
                    { this.props.children }
                </Modal.Body>
                <Modal.Footer>
                    <Button bsStyle={this.props.bsStyle || 'default'} className="popupFooterButton" onClick={e=>{this.props.onClose(e);}}>
                        { this.props.acceptButtonText || 'OK' }
                    </Button>
                </Modal.Footer>
            </Modal>
        );
    }
}
