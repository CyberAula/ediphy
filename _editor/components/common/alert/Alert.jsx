import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button, Modal } from 'react-bootstrap';
import i18n from 'i18next';
/** *
 * Custom alert component
 * @example  <Alert show={this.state.showAlert}      | true : displays alert / false: hides alert
 *                   className="myAlert"             | custom css class (default: popupAlert)
 *                   bsStyle="default"               | bootstrap style (default: default)
 *                   hasHeader                       | whether modal has a header (default: false)
 *                   backdrop                        | backdrop value (same as bootstrap modal) (default: true)
 *                   title="Alert"                   | title displayed in the header (if it has one) (default: empty)
 *                   closeButton                     | close button (x) in the header (if it has one) (default: false)
 *                   acceptButtonText="Accept"       | text displayed in the accept button (default: "OK")
 *                   cancelButton                    | whether there is a cancel button or not (default: false)
 *                   cancelButtonText="Cancel"       | text displayed in the cancel button (if it has one) (default: "Cancel")
 *                   onClose={(bool)=>{...}}>        | callback for when the alert is closed. bool is true when accept button is clicked, false otherwise
 *                        {children}                 | content of the alert. !!!!!! if there are controlled items (e.g. inputs) they need to be handled on the parent component
 *  </Alert>
 */
export default class Alert extends Component {
    /**
     * Constructor
     * @param props React Component props
     */
    constructor(props) {
        super(props);
    }

    /**
     * Renders React component
     * @returns {code}
     */
    render() {
        return(
            <Modal id="alertModal" className={this.props.className || 'popupAlert'}
                onKeyUp={(e)=>{
                    if (e.keyCode === 13) {
                        this.props.onClose(true);
                    }
                }}
                backdrop={this.props.backdrop === undefined ? true : this.props.backdrop}
                show={this.props.show}
                onHide={e=>{this.props.onClose(false);}}>
                { this.props.hasHeader ?
                    (<Modal.Header closeButton={this.props.closeButton}>
                        { this.props.title }
                    </Modal.Header>) :
                    (null)}
                <Modal.Body>
                    { this.props.children }
                </Modal.Body>
                <Modal.Footer>
                    { this.props.cancelButton ? (
                        <Button bsStyle={this.props.bsStyle || 'default'}
                            className="popupFooterButton"
                            onClick={e=>{this.props.onClose(false);}}>
                            { this.props.cancelButtonText || i18n.t('messages.Cancel') }
                        </Button>) : null }
                    <Button bsStyle={this.props.bsStyle || 'primary'}
                        className="popupFooterButton"
                        onClick={e=>{this.props.onClose(true);}}>
                        { this.props.acceptButtonText || i18n.t('messages.OK') }
                    </Button>
                </Modal.Footer>
            </Modal>
        );
    }

}

Alert.propTypes = {
    /**
     * Indica si se muestra o se oculta la alerta
     */
    show: PropTypes.bool,
    /**
     * Clases CSS para aplicar al modal de la alerta
     */
    className: PropTypes.string,
    /**
     * Estilo de Bootstrap para los botones de la alerta
     */
    bsStyle: PropTypes.oneOf(["success", "warning", "danger", "info", "default", "primary", "link"]),
    /**
     * Indica si el modal de la alerta debe tener barra de encabezado
     */
    hasHeader: PropTypes.bool,
    /**
     * Indica si se debe incluir un fondo negro semitransparente detrás de la alerta (true). Si se especifica `static`, no se ocultará la alerta al hacer click.
     */
    backdrop: PropTypes.oneOf(['static', true, false]),
    /**
     * Título para el encabezado. Sólo se mostrará si la propedad `hasHeader` es `true`. Puede ser una cadena de texto o un componente de React.
     */
    title: PropTypes.any,
    /**
     * Indica si se debe incluir un botón de cerrar en el encabezado. Sólo se mostrará si la propedad `hasHeader` es `true`
     */
    closeButton: PropTypes.bool,
    /**
     * Texto del botón de aceptar
     */
    acceptButtonText: PropTypes.string,
    /**
     * Indica si se incluye un botón de cancelar
     */
    cancelButton: PropTypes.bool,
    /**
     * Texto del botón de cancelar (si lo hay)
     */
    cancelButtonText: PropTypes.string,
    /**
     * Cierra la alerta
     */
    onClose: PropTypes.func.isRequired,
    /**
     * Componentes que forman el interior de la alerta. Si son componentes controlados, debe gestionarse su estado desde el componente padre.
     */
    children: PropTypes.any,
};
