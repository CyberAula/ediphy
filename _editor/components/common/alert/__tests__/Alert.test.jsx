import React from 'react';
import expect from 'expect';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Alert from '../Alert';

const mocki18n = 'i18n string';

jest.mock('i18next', () => ({
    t: jest.fn(() => mocki18n).mockName('i18n.t'),
}));

beforeEach(() => {
    // Forget about any mock function being called previously
    jest.clearAllMocks();
});

const mockFunctions = {
    onClose: jest.fn().mockName('onClose'),
};

describe('Alert', () => {
    it('should render with required props', () => {
        const component = shallow(<Alert {...mockFunctions} />);

        expect(shallowToJson(component)).toMatchSnapshot();
    });

    it('should modify the Modal\'s class when the className is set', () => {
        const mockClass = 'ALERT_CLASS';
        const component = shallow(<Alert {...mockFunctions} className={mockClass} />);

        expect(component.find('#alertModal').hasClass(mockClass)).toBeTruthy();
    });

    it('should set the corresponding backdrop when set explictly', () => {
        const component = shallow(<Alert {...mockFunctions} />);

        [true, false, 'static'].map(value => {
            component.setProps({ backdrop: value });

            expect(component.find('#alertModal').prop('backdrop')).toBe(value);
        });
    });

    it('should display both the header and title when specified in the props', () => {
        const mockTitle = 'MOCK_TITLE';
        const component = shallow(<Alert {...mockFunctions}
            hasHeader
            title={mockTitle} />);

        expect(component.find('Modal ModalHeader')).toHaveLength(1);
        expect(component.find('Modal ModalHeader').children().text()).toBe(mockTitle);
    });

    it('should display the cancel button in the footer when specified in the props', () => {
        const mockCancelText = 'MOCK_CANCEL';
        const component = shallow(<Alert {...mockFunctions}
            cancelButton
            cancelButtonText={mockCancelText} />);

        expect(component.find('Modal ModalFooter Button[name="cancelButton"]')).toHaveLength(1);
    });

    it('should set the Bootstrap button style as specified in the props, or use the defaults otherwise', () => {
        const component = shallow(<Alert {...mockFunctions} cancelButton />);

        [undefined, 'success', 'warning', 'danger', 'info', 'default', 'primary', 'link'].map(value => {
            component.setProps({ bsStyle: value });

            expect(component.find('Modal ModalFooter Button[name="cancelButton"]').prop('bsStyle')).toBe(value || 'default');
            expect(component.find('Modal ModalFooter Button[name="okButton"]').prop('bsStyle')).toBe(value || 'primary');
        });
    });

    it('should set the buttons\' inner text as specified in props, or use the defaults otherwise', () => {
        const component = shallow(<Alert {...mockFunctions} cancelButton />);

        [undefined, mocki18n].map(value => {
            component.setProps({
                cancelButtonText: value,
                acceptButtonText: value,
            });

            expect(component.find('Modal ModalFooter Button[name="cancelButton"]').children().text()).toBe(value || mocki18n);
            expect(component.find('Modal ModalFooter Button[name="okButton"]').children().text()).toBe(value || mocki18n);
        });
    });

    it('should close accepting the dialog when ENTER is hit', () => {
        const component = shallow(<Alert {...mockFunctions} />);

        component.simulate('keyUp', { keyCode: 13 });

        expect(mockFunctions.onClose).toBeCalledWith(true);
    });

    it('shouldn\'t do anything when any other key is pressed', () => {
        const component = shallow(<Alert {...mockFunctions} />);

        component.simulate('keyUp', { keyCode: 42 });

        expect(mockFunctions.onClose).not.toBeCalled();
    });

    it('should close rejecting the dialog when onHide is triggered', () => {
        // react-bootstrap automatically calls onHide when the
        // closeButton/backdrop (if present/non-static) is clicked

        const component = shallow(<Alert {...mockFunctions} />);

        component.find('#alertModal').prop('onHide')();

        expect(mockFunctions.onClose).toBeCalledWith(false);
    });

    it('should close rejecting the dialog when the "cancel" button is clicked', () => {
        const component = shallow(<Alert {...mockFunctions} cancelButton />);

        component.find('Modal ModalFooter Button[name="cancelButton"]').simulate('click');

        expect(mockFunctions.onClose).toBeCalledWith(false);
    });

    it('should close accepting the dialog when the "ok" button is clicked', () => {
        const component = shallow(<Alert {...mockFunctions} cancelButton />);

        component.find('Modal ModalFooter Button[name="okButton"]').simulate('click');

        expect(mockFunctions.onClose).toBeCalledWith(true);
    });
});
