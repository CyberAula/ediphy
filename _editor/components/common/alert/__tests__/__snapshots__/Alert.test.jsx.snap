// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should render with required props 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  className="popupAlert"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  id="alertModal"
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[Function]}
  onKeyUp={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={false}
>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  />
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      className="popupFooterButton"
      disabled={false}
      name="okButton"
      onClick={[Function]}
    >
      i18n string
    </Button>
  </ModalFooter>
</Modal>
`;
